Source/Lib/Codec/EbMalloc.h

```diff
@@ -0,0 +1,206 @@
+    EB_PTR_TYPE_TOTAL,
+} EB_PTRType;
+
+#ifdef DEBUG_MEMORY_USAGE
+void EbAddMemEntry(void* ptr,  EB_PTRType type, size_t count, const char* file, uint32_t line);
+void EbRemoveMemEntry(void* ptr, EB_PTRType type);
+
+#define EB_ADD_MEM_ENTRY(p, type, count) \
+    EbAddMemEntry(p, type, count, __FILE__, __LINE__)
+
+#define EB_REMOVE_MEM_ENTRY(p, type) \
+    EbRemoveMemEntry(p, type);
+
+#else
+
+#define EB_ADD_MEM_ENTRY(p, type, count)
+#define EB_REMOVE_MEM_ENTRY(p, type)
+
+#endif //DEBUG_MEMORY_USAGE
+
+#define EB_NO_THROW_ADD_MEM(p, size, type) \
+    do { \
+        if (!p) { \
+            fprintf(stderr,"allocate memory failed, at %s, L%d\n", __FILE__, __LINE__); \
+        } else { \
+            EB_ADD_MEM_ENTRY(p, type, size); \
+        } \
+    } while (0)
+
+#define EB_CHECK_MEM(p) \
+    do { \
+        if (!p) {\
+            return EB_ErrorInsufficientResources; \
+        } \
+    } while (0)
+
+#define EB_ADD_MEM(p, size, type) \
+    do { \
+        EB_NO_THROW_ADD_MEM(p, size, type); \
+        EB_CHECK_MEM(p); \
+    } while (0)
+
+#define EB_NO_THROW_MALLOC(pointer, size) \
+    do { \
+        void* p = malloc(size); \
+        EB_NO_THROW_ADD_MEM(p, size, EB_N_PTR); \
+        *(void**)&(pointer) = p; \
+    } while (0)
+
+#define EB_MALLOC(pointer, size) \
+    do { \
+        EB_NO_THROW_MALLOC(pointer, size); \
+        EB_CHECK_MEM(pointer); \
+    } while (0)
+
+#define EB_NO_THROW_CALLOC(pointer, count, size) \
+    do { \
+        void* p = calloc(count, size); \
+        EB_NO_THROW_ADD_MEM(p, count * size, EB_C_PTR); \
+        *(void**)&(pointer) = p; \
+    } while (0)
+
+#define EB_CALLOC(pointer, count, size) \
+    do { \
+        EB_NO_THROW_CALLOC(pointer, count, size); \
+        EB_CHECK_MEM(pointer); \
+    } while (0)
+
+#define EB_FREE(pointer) \
+    do {\
+        free(pointer); \
+        EB_REMOVE_MEM_ENTRY(pointer, EB_N_PTR); \
+        pointer = NULL; \
+    } while (0)
+
+
+#define EB_MALLOC_ARRAY(pa, count) \
+    do {\
+        size_t size = sizeof(*(pa)); \
+        EB_MALLOC(pa, (count)*size); \
+    } while (0)
+
+#define EB_CALLOC_ARRAY(pa, count) \
+    do {\
+        size_t size = sizeof(*(pa)); \
+        EB_CALLOC(pa, count, size); \
+    } while (0)
+
+#define EB_FREE_ARRAY(pa) \
+    EB_FREE(pa);
+
+
+#define EB_ALLOC_PTR_ARRAY(pa, count) \
+    do {\
+        size_t size = sizeof(*(pa)); \
+        EB_CALLOC(pa, count, size); \
+    } while (0)
+
+#define EB_FREE_PTR_ARRAY(pa, count) \
+    do {\
+        if (pa) { \
+            uint32_t i; \
+            for (i = 0; i < count; i++) { \
+                EB_FREE(pa[i]); \
+            } \
+            EB_FREE(pa); \
+        } \
+    } while (0)
+
+
+#define EB_MALLOC_2D(p2d, width, height) \
+    do {\
+        EB_MALLOC_ARRAY(p2d, width); \
+        EB_MALLOC_ARRAY(p2d[0], width * height); \
+        for (uint32_t w = 1; w < width; w++) { \
+            p2d[w] = p2d[0] + w * height; \
```

Comment:
This a tread off, if we do not do this, we will have huge malloc times. suppose we have a 1920x1080  it will allocate  2M times.
If we do think this a problem, at least we need to allocate a line for each time.
The malloc times will drop to 2M to 1080.

Reply:
1. According to malloc [man](http://man7.org/linux/man-pages/man3/malloc.3.html) (section NOTES) and the MM related [Q&A](https://landley.net/writing/memory-faq.txt), "_The Linux kernel uses lazy (on-demand) allocation of physical pages, deferring the allocation until necessary and avoiding allocating physical pages which will never actually be used._" Windows should have the similar allocation strategy. malloc wouldn't take much time, except for setting up the corresponding page table. And page fault handler will allocate physical memory and map the pages with virtuall address, once you really need to read, write or map it.
2. So there wouldn't be big time consuming difference between: 1) malloc memories for each entry of the pointer array; AND 2) as implemented here, malloc all the memory with 1 time, then assign each entry of the pointer array with the equally spaced virtural addresses.
3. With current EB_MALLOC_2D implementation, free any entry (except for p2d[0]) of the pointer array will fail, when those memories are not needed any more. All of them needs to wait for freeing p2d[0] to be freed.
4. As I mentioned before, such implementation requires that the allocator must have continuous (no hole) linear space with that "big" size, although the allocation failure almost wouldn't happen. :-)

Reply:
1. bind phisical memory is not cost one, How to find the right size for malloc and how to split large memory to meet our requirement will use most of cpu time. Usally, if you do not do any optimization, for a decoder, ~20~30 cpu time will used by malloc. Encoder may less than this, but it still a noticalbe one.
3.  we never remove any entries, svt alwasy allocate system at start, and free it at end.
4. it's just virtual address not physical, For 64 bits system, it usally has 48bits address space,  whichi is 281TB, so you always can find virtual addreess for our allocation.

Maybe we can do some profile to see the time of multipile malloc vs. one malloc.

Reply:
1 & 2. I noticed that in user space, multiple mallocs (for the same total size) take more time due to syscalls (brk or mmap) and allocation algorithms done in glibc or kernel (Linux slab) allocator. So in theory, the EB_MALLOC_2D could consume less time, especially when width or height is very big.
3. It can't address the separate free issue, unless there is additional resource management such as refcount.
4. Yes, I was talking about virtual address (space).